pyro/ops/studentt.py

```diff
@@ -0,0 +1,324 @@
+
+import pyro
+from pyro.distributions.util import broadcast_shape
+from pyro.ops.gamma_gaussian import Gamma, GammaGaussian, gamma_and_mvn_to_gamma_gaussian
+
+
+def _moment_matching(st, new_df, order=1):
+    """
+    Approximates a StudentT by another one with different degree of freedom
+    using absolute central moment matching.
+    """
+    # Ref: Moments of Student's t-distribution: A Unified Approach,
+    # J. Lars Kirkby, Dang Nguyen, Duy Nguyen https://arxiv.org/abs/1912.01607
+    # We want to find s such that
+    #   St(df, m, P) ~ St(new_df, m', sP)
+    # Matching the mean gives us m' = m.
+    # i.e. St(df, 0, I) ~ St(new_df, 0, sI)
+    # Given a moment with order k1, k2,... such that sum(k_i) = 1,
+    # we will use the absolute moment formula in Theorem 3 of the above reference
+    # and match the corresponding moments of two distributions.
+    #
+    # Given s, we need to adjust beta, info_vec, precision for moment matching.
+    # We will use the results of moment matching to take the sum of two uncorrelated student-t,
+    # so it is better to have new rate = new concentration = new_df / 2.
+    # We know that (see GammaGaussian.compound() method) the precision of
+    #   StudentT(alpha, beta, info_vec, precision)
+    # is concentration / rate * precision, where rate = beta - 0.5 uPu.
+    # So we want to find P' = s'P such that
+    #   s' = concentration / rate * s.
+    assert (order < new_df).all()
+    half_mask_sum = 0.5 * st.mask.sum()
+    concentration = st.joint.alpha - half_mask_sum + 1
+    new_concentration = 0.5 * new_df
+    new_alpha = new_concentration + half_mask_sum - 1
+    chol_P = st.joint.precision[..., st.mask, :][..., st.mask].cholesky()
+    chol_P_u = st.joint.info_vec[..., st.mask].unsqueeze(-1).triangular_solve(chol_P, upper=False).solution.squeeze(-1)
+    half_u_P_u = 0.5 * chol_P_u.pow(2).sum(-1)
+    rate = st.joint.beta - half_u_P_u
+    half_order = 0.5 * order
+    s = (new_concentration / rate) * torch.exp((
+        torch.lgamma(new_concentration - half_order) - torch.lgamma(new_concentration)
+        - torch.lgamma(concentration - half_order) + torch.lgamma(concentration)) / half_order)
+    new_beta = new_concentration + s * half_u_P_u
+    # adjust log_normalizer
+    info_vec = st.joint.info_vec * s.unsqueeze(-1)
+    precision = st.joint.precision * s.unsqueeze(-1).unsqueeze(-1)
+    log_normalizer = st.joint.log_normalizer + Gamma(0., concentration, rate).logsumexp() - \
+        Gamma(0., new_concentration, new_concentration).logsumexp() + half_mask_sum * s.log()
+    return StudentT(log_normalizer, info_vec, precision, new_alpha, new_beta, st.mask)
+
+
+class StudentT(GammaGaussian):
+    """
+    Non-normalized StudentT distribution:
+
+        StudentT(x) ~ Integral(p(x, s), s),
+
+    where
+
+        p(x, s) = GammaGaussian(log_normalizer, info_vec, precision, alpha, beta).
+
+    Note that different StudentT instances can have the same log density.
```

Comment:
Is this because `Studentt(l,i,p,a,b) = StudentT(l,i*t,p*t,a,b/t)` for any `t>0`? If so could you add that to this comment?

Reply:
Yes, it is (modulo alpha, beta reparameterization). I'll clarify it.